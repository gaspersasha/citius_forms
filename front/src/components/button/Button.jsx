import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { Heart } from '~assets/svg';
import s from './button.module.sass';

/**
 * Button component
 * Allows to configure appearance
 * & behavior (<a/> or <button/> element) through props
 */
const Button = (props) => {
  const { children, className, href, styleType, selected, ...rest } = props;

  // render link element if href prop is provided
  const Component = href ? 'a' : 'button';

  // composing selectors from provided props
  // attribute 'disabled' will apply styles for disabled type button
  const composedClassName = cn(
    s.buttonComponent,
    {
      [s.primary]: styleType === 'primary',
      [s.secondary]: styleType === 'secondary',
      [s.like]: styleType === 'like',
      [s[`${styleType}-selected`]]: selected,
      [s.cancel]: styleType === 'cancel',
    },
    className
  );

  return (
    <Component href={href || null} className={composedClassName} {...rest}>
      {children}
      {!children && styleType === 'like' && <Heart />}
    </Component>
  );
};

Button.propTypes = {
  // whatever will be rendered insine the component
  children: PropTypes.node,
  // component's style type
  styleType: PropTypes.oneOf(['primary', 'secondary', 'like', 'cancel']),
  // css selector to apply to the root element
  className: PropTypes.string,
  // url to redirect on component's click
  // if provided, component will behave like a link
  href: PropTypes.string,
  // makes secondary/like type button selected i.e. filled
  selected: PropTypes.bool,
  // button's disabled state, will apply styles for disabled type button
  disabled: PropTypes.bool,
  // callback for the button's onClick event
  onClick: PropTypes.func,
};

Button.defaultProps = {
  children: null,
  className: '',
  href: '',
  selected: false,
  disabled: false,
  onClick: null,
  styleType: 'primary',
};

export default Button;
