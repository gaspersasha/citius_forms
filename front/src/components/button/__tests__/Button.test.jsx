import React from 'react';
import Button from '../Button';

describe('<Button />', () => {
  let wrapper;

  it('rendered without errors', () => {
    wrapper = shallow(<Button>Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with default parameters', () => {
    wrapper = shallow(<Button>Test</Button>);

    expect(wrapper.find('button').prop('href')).toBe(null);
    expect(wrapper.find('button').prop('disabled')).toBe(false);
    expect(wrapper.find('button').prop('className')).toEqual(
      'buttonComponent primary'
    );
  });

  it('renders children properly', () => {
    wrapper = shallow(
      <Button>
        <span>Test</span>
      </Button>
    );

    expect(wrapper.find('button').children().html()).toEqual(
      '<span>Test</span>'
    );
  });

  it('passes className prop', () => {
    wrapper = shallow(<Button className="test">Test</Button>);
    expect(wrapper.find('button').prop('className')).toEqual(
      'buttonComponent primary test'
    );
  });

  it('disabled case', () => {
    const fn = jest.fn();

    wrapper = mount(
      <Button disabled onClick={fn}>
        Test
      </Button>
    );

    wrapper.find('button').simulate('click');
    wrapper.update();

    expect(fn).not.toBeCalled();
    expect(wrapper.find('button').prop('disabled')).toBe(true);
  });

  it('handles provided events', () => {
    const fn = jest.fn();

    wrapper = mount(<Button onClick={fn}>Test</Button>);

    wrapper.find('button').simulate('click');
    wrapper.update();

    expect(fn).toBeCalled();
  });

  it('link component case', () => {
    wrapper = shallow(<Button href="/href">Test</Button>);
    expect(wrapper.find('a').prop('href')).toEqual('/href');
  });

  it.skip('changes appearance', () => {
    wrapper = shallow(
      <Button color="green" txtColor="pink" backgroundType="stroke">
        Test
      </Button>
    );

    expect(wrapper.find('button').prop('className')).toBe(
      'buttonComponent txt-pink shadow-green stroke-green'
    );
  });
});
