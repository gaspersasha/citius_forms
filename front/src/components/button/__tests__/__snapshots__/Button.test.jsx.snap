// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> rendered without errors 1`] = `
<button
  className="buttonComponent primary"
  disabled={false}
  href={null}
  onClick={null}
>
  Test
</button>
`;
